spec/validator.ml

```diff
@@ -1,180 +1,111 @@
-  share : Secp256k1.secret_key;
+  key : FROST.verification_key;
+  me : FROST.Identifier.t local;
+  share : FROST.signing_share local;
 }
 
-type epoch = { id : int; group : group }
+type epoch = { epoch : int; group : group }
 
 type rollover =
   | Collecting_key_gen_commitments of {
       epoch : int;
-      id : group_id;
-      coefficients : Secp256k1.scalar list;
-      commitments : Secp256k1.point list IntMap.t;
-      participants : ParticipantSet.t;
+      group : GroupId.t;
+      participants : AddressSet.t;
+      coefficients : FROST.KeyGen.coefficients local;
+      commitments : FROST.KeyGen.commitments ParticipantMap.t;
       deadline : int;
     }
   | Collecting_key_gen_secret_shares of {
       epoch : int;
-      id : group_id;
-      key : Secp256k1.point;
-      commitments : Secp256k1.point list IntMap.t;
-      shares : Secp256k1.scalar IntMap.t;
-      last_participant : Address.t;
-      participants : ParticipantSet.t;
+      group : GroupId.t;
+      participants : AddressSet.t;
+      key : FROST.verification_key;
+      coefficients : FROST.KeyGen.coefficients local;
+      commitments : FROST.KeyGen.commitments ParticipantMap.t;
+      shares : FROST.KeyGen.secret_share ParticipantMap.t local;
+      last_participant : FROST.Identifier.t;
       deadline : int;
     }
   | Signing_epoch_rollover of { epoch : epoch; message : string }
-  | Staged_epoch of epoch
+  | Staged_epoch of { epoch : epoch }
+  | Skipped of { epoch : int }
 
 type preprocess_group = {
-  nonces : FROST.nonces IntMap.t;
-  pending : FROST.nonces list StringMap.t;
+  nonces : (FROST.nonces * MerkleTree.proof) IntMap.t;
+  pending : (int * FROST.nonces) list option;
 }
 
-type signature_id = string
-
-type sign_status =
-  | Waiting_for_sign_request of { responsible : Address.t }
+type sign_state =
+  | Waiting_for_sign_request of { responsible : FROST.Identifier.t option }
   | Collecting_sign_nonces of {
-      id : signature_id;
-      nonces : FROST.nonces_commitments IntMap.t;
+      id : SignatureId.t;
+      nonces : FROST.nonces_commitments ParticipantMap.t;
     }
   | Collecting_sign_shares of {
-      id : signature_id;
+      id : SignatureId.t;
       root : string;
-      group_commitment : Secp256k1.point;
-      shares : Secp256k1.scalar IntMap.t;
+      group_commitment : FROST.group_commitment;
+      shares : FROST.signature_share ParticipantMap.t;
     }
   | Waiting_for_consensus_attestation of {
-      id : signature_id;
-      responsible : Address.t;
+      id : SignatureId.t;
+      responsible : FROST.Identifier.t option;
     }
 
 type packet =
-  | Epoch_rollover of { epoch : int; rollover : int; group_id : group_id }
+  | Epoch_rollover of {
+      epoch : int;
+      rollover : int;
+      group_key : FROST.verification_key;
+    }
   | Transaction_proposal of { epoch : int; hash : string }
 
-type signing_ceremony = {
-  status : sign_status;
+type signature = {
+  state : sign_state;
```

Comment:
reminder: in the validator code we also use a lot "status", might make sense to align this at some point